### PR TITLE
new task: pointer network for joint ner and re

### PR DESCRIPTION
**Pull request by ArneBinder**
code is mostly taken from https://github.com/ArneBinder/pie-document-level/pull/26

Note: This adds the requirement `transformers = "^4.32.0"`, so that we can import `BartPreTrainedModel` (**breaking**).

Requires: 
 - pytorch-ie>=0.29.5 (https://github.com/ArneBinder/pytorch-ie/pull/392)
 - #24 (for the PIE model)
 - #25 (for metrics)
 - #26 (for `decoder_position_id_pattern`, i.e. replaced positional encodings (RPE))

TODO:
 - [x] taskmodule: add code and get tests passing
 - [x] taskmodule: understand by refactoring (x + 3h + 8h)
 - [x] model: add code and get tests passing (prediction: 8h, true: 5h)
 - [x] model: understand by refactoring (estimation: 16h, true: 3h+?)
    - [x] cleanup ids and get old setup running (true: 3h) 
    - [x] implement simplified version (estimation: 8h, true: 42h):
        - [x] #20 (estimation: 8h, true: 28h)
        - [x] no RPE (estimation: 2h, true: 1h)
        - [x] mask train labels in simple model (estimation: 1h, true: 1h)
        - [x] understand and fix weight decay issues (true: 6h)
        - [x] find a good way what to log when (finally, add `metric_intervals` parameter) (true: 2h)
        - [x] show that training is aligned (true: 4h)
- [x] use `pytorch-ie = ">=0.29.5,<0.30.0"` when released (we require https://github.com/ArneBinder/pytorch-ie/pull/392) - [x] constrained training (true: 4h)
- [x] taskmodule
     - [x] merge `PointerNetworkSpanAndRelationEncoderDecoder` back again into taskmodule and improve structure (true: 10h)
     - [x] there are a lot of missed annotations during tokenization... EDIT: should be fixed (we did not set `return_overflowing_tokens=True`)
     - [x] check if `cmp_src_rel` sorts the relations in a useful manner (note: we encode relations as `tail-head-label`). Edit: we sort first by head, then by tail. We could try the other way, which should be more in sync with the encoding mode, but the current setup seems to work.
     - [x] understand `sanitize_sequence`
     - [x] tests
       - [x] increase coverage
           - [x] add tests for `common` and `pointer_network` modules (true: 2h)
           - [x] add tests for `sanitize_sequence()`
       - [x]  simplify / disentangle existing ones
     - [x] rename (we tackle `end2end_re` with the current taskmodule variant) . Edit: fixed in 3411b298bbc6bb8e5ee5fbeaf0e377d20a89c387
     - [x] generalize metrics
        - [x] add / improve tests
    - [x] use [PrefixConstrainedLogitsProcessor](https://huggingface.co/docs/transformers/internal/generation_utils#transformers.PrefixConstrainedLogitsProcessor) to constrain the generation (true: 3h + 5h)
 - [x] model: 
   - [x] tests
     - [x] increase coverage
     - [x] add `test_pointer_head.py` (`PointerHead`)
     - [x] improve `test_bart_as_pointer_network` (`BartAsPointerNetwork`)
     - [x] remove `test_simple_generative_pointer.py`
     - [x] remove `test_simple_generative_pointer_predict.py`
     - [x] check if fixture data is unused and delete it if that is the case
     - [x] ~use mock models?~ Edit: ~use [sshleifer/tiny-mbart](https://huggingface.co/sshleifer/tiny-mbart)~ use [sshleifer/bart-tiny-random](https://huggingface.co/sshleifer/bart-tiny-random)
    - [x] improve weight decay parametrization (currently we use a single parameter `layernorm_decay` to set "encoder only layernorm parameters"). Edit: fixed in b8da7e31509796a2e09f83ba5e35adb0952d1f26
   - [x] fix metric logging warning. Edit: fixed in 007064906e55b827394bd74d70c1eaaafa470f25
   - [x] move original model into separate branch. Edit: done, see https://github.com/ArneBinder/pie-modules/tree/original_pointer_network
   - [x] parametrize if to use predictions or model output to calculate metrics (the latter should be much faster, but overestimating). Implemented in 6b01b1b7cb678eacbadee3a34bca3c3e49bcc70a
   - [x] implement RPE. Edit: finalized in d5facf9793fe413bd48f8e02b0846e4b1761a4d7
   - [x] generalize PIE model to `GenerativeModel`. Edit: see #24
   - [x] outsource `BartModelWithDecoderPositionIds`, see #26
   - ~move `prepare_decoder_position_ids` to `BartAsPointerNetwork`~ Edit: no, this is fine
   - [x] ~use `special_tokens_mask` in `PointerHead`~ Edit: no, just mask eos, bos, and pad positions in offset scores
   

Follow-up:  
 - adapt other base models, see [`transformers.AutoModelForSeq2SeqLM`](https://github.com/huggingface/transformers/blob/v4.36.1/src/transformers/models/auto/modeling_auto.py#L683-L712) for candidates
   - e.g. `LongT5ForConditionalGeneration` ([pszemraj/long-t5-tglobal-base-sci-simplify](https://huggingface.co/pszemraj/long-t5-tglobal-base-sci-simplify))
- maybe: use `BucketSampler`, see https://github.com/Lightning-AI/lightning/pull/13640#issuecomment-1199032224

### CODE REVIEW

**Comment by ArneBinder** [comment]
It looks like the training is now aligned. When we replace the weights of the original model with the ones from the simple model, we get [the following](https://github.com/ArneBinder/pie-modules/blob/710de7bc46c394d7c96cccbdf795030c36fdb3d5/tests/models/test_pointer_network_model.py#L835-L847):
```python
# check the numbers
assert {layer_name: len(anns) for layer_name, anns in fp.items()} == {
    "labeled_spans": 101,
    "binary_relations": 63,
}
assert {layer_name: len(anns) for layer_name, anns in fn.items()} == {
    "labeled_spans": 93,
    "binary_relations": 73,
}
assert {layer_name: len(anns) for layer_name, anns in tp.items()} == {
    "labeled_spans": 41,
    "binary_relations": 9,
}
```

which is quite close to [the result from the not-modified original model](https://github.com/ArneBinder/pie-modules/blob/710de7bc46c394d7c96cccbdf795030c36fdb3d5/tests/models/test_pointer_network_model.py#L835-L847):
```python
# check the numbers
assert {layer_name: len(anns) for layer_name, anns in fp.items()} == {
    "labeled_spans": 126,
    "binary_relations": 72,
}
assert {layer_name: len(anns) for layer_name, anns in fn.items()} == {
    "labeled_spans": 88,
    "binary_relations": 73,
}
assert {layer_name: len(anns) for layer_name, anns in tp.items()} == {
    "labeled_spans": 46,
    "binary_relations": 9,
}
```

The result for the simple model we took the weights from [is much worse](https://github.com/ArneBinder/pie-modules/blob/710de7bc46c394d7c96cccbdf795030c36fdb3d5/tests/models/test_simple_pointer_network.py#L444-L456):
```python
# check the numbers
assert {layer_name: len(anns) for layer_name, anns in fp.items()} == {
    "labeled_spans": 108,
    "binary_relations": 65,
}
assert {layer_name: len(anns) for layer_name, anns in fn.items()} == {
    "labeled_spans": 106,
    "binary_relations": 78,
}
assert {layer_name: len(anns) for layer_name, anns in tp.items()} == {
    "labeled_spans": 29,
    "binary_relations": 4,
}
```
